File: 212. Word Search II.py
class TrieNode:
    def __init__(self):
        self.child = {}
        self.isEnd = False
        
class Trie:
    def __init__(self):
        self.root = TrieNode()
        
    def add(self, word):
        l = len(word)
        curr = self.root
        for i in range(l):
            if word[i] not in curr.child:
                curr.child[word[i]] = TrieNode()
            curr = curr.child[word[i]]
        curr.isEnd = True
        return

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        dir = [(1,0), (0,1), (-1,0), (0,-1)]
        # build Trie
        T = Trie()
        for word in words:
            T.add(word)
        res = set()
        
        def check(board, i, j, node, ls):
            if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
                return
            c = board[i][j]
            board[i][j] = '#'
            if c in node.child:
                if node.child[c].isEnd:
                    res.add(''.join(ls+[c]))
                for di, dj in dir:
                    check(board, i+di, j+dj, node.child[c], ls + [c])
            board[i][j] = c
            
        if not board:
            return res
        for i in range(len(board)):
            for j in range(len(board[0])):
                check(board, i, j, T.root, [])
        return res

File: 212. Word Search II.java
class TrieNode {
    boolean isEnd;
    TrieNode[] child;
    public TrieNode() {
        isEnd = false;
        child = new TrieNode[26];
    }
}

class Trie {
    TrieNode root;
    public Trie() {
        root = new TrieNode();
    }
    void add(String word) {
        TrieNode curr = root;
        for (int i=0; i<word.length(); i++) {
            char c = word.charAt(i);
            if (curr.child[c-'a'] == null) {
                curr.child[c-'a'] = new TrieNode();
            }
            curr = curr.child[c-'a'];
        }
        curr.isEnd = true;
    }
}

class Solution {
    char[][] board;
    int row, col;
    Trie t;
    int maxlen;
    public List<String> findWords(char[][] board, String[] words) {
        if (board == null || board.length==0 || board[0].length==0 || words==null || words.length==0) {
            return new ArrayList<>();
        }
        this.board = board;
        row = board.length;
        col = board[0].length;
        t = new Trie();
        maxlen = 0;
        HashSet<String> ans = new HashSet<>();
        for (String word : words) {
            t.add(word);
            maxlen = Math.max(word.length(), maxlen);
        }
        for (int i=0; i<row; i++) {
            for (int j=0; j<col; j++) {
                check(i, j, new StringBuilder(), t.root, ans);
            }
        }
        return new ArrayList<>(ans);
    }
    private void check(int i, int j, StringBuilder sb, TrieNode curr, HashSet<String> ans) {
        if (sb.length() > maxlen) {
            return;
        }
        if (board[i][j] != '#' && curr.child[board[i][j]-'a']!=null) {
            sb.append(board[i][j]);
            if (curr.child[board[i][j]-'a'].isEnd) {
                ans.add(sb.toString());
            }
            char letter = board[i][j];
            board[i][j] = '#';
            if (i>0 && board[i-1][j] != '#') {
                check(i-1, j, sb, curr.child[letter-'a'], ans);
            }
            if (i<row-1 && board[i+1][j] != '#') {
                check(i+1, j, sb, curr.child[letter-'a'], ans);
            }
            if (j>0 && board[i][j-1] != '#') {
                check(i, j-1, sb, curr.child[letter-'a'], ans);
            }
            if (j<col-1 && board[i][j+1] != '#') {
                check(i, j+1, sb, curr.child[letter-'a'], ans);
            }
            sb.deleteCharAt(sb.length()-1);
            board[i][j] = letter;
        }
    }
}